src/pymor/reductors/aaa.py

```diff
@@ -0,0 +1,417 @@
+    ----------
+    samples
+        Tensor of samples (see :class:`pAAAReductor`).
+    svs
+        List of sampling values (see :class:`pAAAReductor`).
+    itpl_part
+        Nested list such that `itpl_part[i]` is a list of indices for interpolated
+        sampling values in `svs[i]`.
+
+    Returns
+    -------
+    L
+        (Parametric) Loewner matrix based only on LS partition.
+    """
+    d = len(samples.shape)
+    ls_part = [sorted(set(range(len(s))) - set(p)) for p, s in zip(itpl_part, svs)]
+
+    s0 = svs[0]
+    s = s0[itpl_part[0]]
+    sh = s0[ls_part[0]]
+    sd = sh[:, np.newaxis] - s
+    sdpd = sd
+    for i in range(1, d):
+        p0 = svs[i]
+        p = p0[itpl_part[i]]
+        ph = p0[ls_part[i]]
+        pd = ph[:, np.newaxis] - p
+        sdpd = np.kron(sdpd, pd)
+    samples0 = samples[np.ix_(*itpl_part)].reshape(-1)
+    samples1 = samples[np.ix_(*ls_part)].reshape(-1, 1)
+    samplesd = samples1 - samples0
+
+    return samplesd / sdpd
+
+
+def full_nd_loewner(samples, svs, itpl_part):
+    """Compute higher-dimensional Loewner matrix using all combinations of partitions.
+
+    .. note::
+       For non-parametric data this is simply the regular Loewner matrix.
+
+    Parameters
+    ----------
+    samples
+        Tensor of samples (see :class:`pAAAReductor`).
+    svs
+        List of sampling values (see :class:`pAAAReductor`).
+    itpl_part
+        Nested list such that `itpl_part[i]` is a list of indices for interpolated
+        sampling values in `svs[i]`.
+
+    Returns
+    -------
+    L
+        (Parametric) Loewner matrix based on all combinations of partitions.
+    """
+    L = nd_loewner(samples, svs, itpl_part)
+    range_S = range(len(svs))
+
+    # consider all combinations of variables coming from interpolation vs LS partition
+    for i in itertools.product(*([0, 1] for _ in range_S)):
+
+        # skip cases corresponding to all interpolated or all LS fit
+        if i == len(svs) * (0,) or i == len(svs) * (1,):
+            continue
+
+        for j in itertools.product(*(itpl_part[k] for k in range_S if i[k] == 0)):
+            l_j = list(j)
+            for ii in range(len(i)):
+                if i[ii] == 1:
+                    l_j.insert(ii, slice(None))
+            samples0 = samples[tuple(l_j)]
+            svs0 = [svs[k] for k in range_S if i[k] == 1]
+            itpl_part0 = [itpl_part[k] for k in range_S if i[k] == 1]
+            T_mat = 1
+            for k in range_S:
+                if i[k] == 1:
+                    T_new = np.eye(len(itpl_part[k]))
+                else:
+                    idx = np.where(np.array(itpl_part[k]) == l_j[k])[0][0]
+                    T_new = np.eye(1, len(itpl_part[k]), idx)
+                T_mat = np.kron(T_mat, T_new)
+            LL = nd_loewner(samples0, svs0, itpl_part0)
+            L = np.vstack([L, LL @ T_mat])
+    return L
+
+
+def make_bary_func(itpl_nodes, itpl_vals, coefs, removable_singularity_tol=1e-14):
```

Comment:
I suppose we could have something like `BarycentricFunction` as a subclass of `TransferFunction` (and similar for pole-residue form, etc.). But, we can discuss this first in #1794 and maybe implement it in another PR.

Reply:
> I suppose we could have something like `BarycentricFunction` as a subclass of `TransferFunction` (and similar for pole-residue form, etc.). But, we can discuss this first in #1794 and maybe implement it in another PR.

That sounds like a good idea. And yes, I will try writing the entire formula here (it is a bit complicated, I will see how it looks).

Reply:
I just wrote the two-variable formula there. The general formula is just to ugly and won't help understanding what is going on in my opinion. If you think there is value in having the general (i.e., arbitrary number of parameters) formula let me know.

Reply:
Agreed. It looks great already.